clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -128,11 +128,12 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                     long timeNs = Math.max(0L, endWaitNs - startWaitNs);
                     this.waitTime.record(timeNs, time.milliseconds());
 
-                    if (waitingTimeElapsed)
+                    if (waitingTimeElapsed) {
+                        this.waiters.removeLast();
```

Comment:
It seems to me that `waiters.remove(moreMemory)` would work. As I said in a previous comment, it would be great if the unit test showed that `waiters.removeLast` doesn't work for all cases.